### PR TITLE
stdlib: Improve `Never` documentation

### PR DESCRIPTION
**Pull request by bradleymackey**
This makes the documentation for `Never` more user-facing with
additional examples of its usage.

Places additional emphasis on types that are impossible to construct,
as real-world usage of Swift makes more heavy use of this
particular facet of `Never` (e.g. Combine) as compared to "not returning
normally".

### CODE REVIEW

File: stdlib/public/core/Policy.swift [hunk 1]
```diff
@@ -15,13 +15,33 @@
 //===----------------------------------------------------------------------===//
 // Standardized uninhabited type
 //===----------------------------------------------------------------------===//
-/// The return type of functions that do not return normally, that is, a type
-/// with no values.
+/// The return type of functions that cannot return a value.
+///
+/// As an uninhabited type, `Never` has no values and cannot be constructed
```
Comment (CodaFi) [inline stdlib/public/core/Policy.swift:20]: We should be careful to define the term `uninhabited` as that's type-theoretic jargon.

File: stdlib/public/core/Policy.swift [hunk 2]
```diff
@@ -15,13 +15,33 @@
 //===----------------------------------------------------------------------===//
 // Standardized uninhabited type
 //===----------------------------------------------------------------------===//
-/// The return type of functions that do not return normally, that is, a type
-/// with no values.
+/// The return type of functions that cannot return a value.
+///
+/// As an uninhabited type, `Never` has no values and cannot be constructed 
+/// normally, allowing you to represent a state in your program that is
```
Comment (CodaFi) [inline stdlib/public/core/Policy.swift:21]: strike `normally`.

File: stdlib/public/core/Policy.swift [hunk 3]
```diff
@@ -15,13 +15,33 @@
 //===----------------------------------------------------------------------===//
 // Standardized uninhabited type
 //===----------------------------------------------------------------------===//
-/// The return type of functions that do not return normally, that is, a type
-/// with no values.
+/// The return type of functions that cannot return a value.
+///
+/// As an uninhabited type, `Never` has no values and cannot be constructed 
+/// normally, allowing you to represent a state in your program that is 
+/// impossible to reach during execution.
```
Comment (CodaFi) [inline stdlib/public/core/Policy.swift:22]:
```suggestion
/// impossible to reach during its execution.
```

File: stdlib/public/core/Policy.swift [hunk 4]
```diff
@@ -15,13 +15,33 @@
...
+///     // The `.failure` case cannot be reached because it cannot be 
+///     // constructed. Therefore, Swift allows us to elide it.
+///     let favouriteNumber: Result<Int, Never> = .success(42)
+///     switch favouriteNumber {
+///     case .success(let value):
+///         print("My favourite number is", value)
+///     }
 ///
 /// Use `Never` as the return type when declaring a closure, function, or
-/// method that unconditionally throws an error, traps, or otherwise does
```
Comment (CodaFi) [inline stdlib/public/core/Policy.swift:22]: Leave this as is. There are other domain-specific ways for programs to halt that this covers.

**Comment by CodaFi** [comment]
Some nits in line. Nothing too serious - take these as you wish.

File: stdlib/public/core/Policy.swift (hunk 1 above)
Comment (bradleymackey) [reply stdlib/public/core/Policy.swift]: Rephrased to remove this word.

File: stdlib/public/core/Policy.swift [hunk 5]
```diff
@@ -15,8 +15,26 @@
 //===----------------------------------------------------------------------===//
 // Standardized uninhabited type
 //===----------------------------------------------------------------------===//
-/// The return type of functions that do not return normally, that is, a type
-/// with no values.
+/// The return type of functions that cannot return a value.
```
Comment (benrimmington) [inline stdlib/public/core/Policy.swift:18]: *The Swift Programming Language* book calls them "[nonreturning][]" functions.

[nonreturning]: <https://docs.swift.org/swift-book/ReferenceManual/Declarations.html#ID551>

File: stdlib/public/core/Policy.swift [hunk 6]
```diff
@@ -15,8 +15,26 @@
...
+/// `Never` is a type with no values and cannot be constructed, allowing you
+/// to represent a state in your program that is impossible to reach during
+/// its execution.
+///
+///     // An infinite loop will never terminate.
+///     func forever() -> Never {
+///         while true {
+///             print("I will print forever and will not return.")
+///         }
+///     }
```
Comment (benrimmington) [inline stdlib/public/core/Policy.swift:29]: Your `forever()` example could move further down, alongside `crashAndBurn()`.

File: stdlib/public/core/Policy.swift [hunk 7]
```diff
@@ -15,8 +15,26 @@
...
+///
+///     // The `.failure` case cannot be reached because it cannot be 
+///     // constructed; this makes the following switch exhaustive.
+///     let favouriteNumber: Result<Int, Never> = .success(42)
+///     switch favouriteNumber {
+///     case .success(let value):
+///         print("My favourite number is", value)
```
Comment (benrimmington) [inline stdlib/public/core/Policy.swift:36]: US English spelling?

```suggestion
///     let favoriteNumber: Result<Int, Never> = .success(42)
///     switch favoriteNumber {
///     case .success(let value):
///         print("My favorite number is", value)
```

File: stdlib/public/core/Policy.swift (hunk 6 above)
Comment (bradleymackey) [reply stdlib/public/core/Policy.swift:30]: Added a snippet about *nonreturning* functions from the Swift Reference, so I've placed it below that. It seems this makes the most sense and separates each example snippet.

File: stdlib/public/core/Policy.swift [hunk 8]
```diff
@@ -15,16 +15,40 @@
...
+///     // constructed; this makes the following switch exhaustive.
+///     let favoriteNumber: Result<Int, Never> = .success(42)
+///     switch favoriteNumber {
+///     case .success(let value):
+///         print("My favorite number is", value)
+///     }
+///
+/// Use `Never` as the return type when declaring a closure, function, or
+/// method that unconditionally throws an error, traps, or otherwise does
+/// not terminate.
```
Comment (benrimmington) [inline stdlib/public/core/Policy.swift:51]: Should this paragraph be removed?

```suggestion
```

---

Should conformances to standard and non-standard protocols be discussed?
For example, the [Conform Never to Identifiable][] pitch uses an empty `switch` statement.

[Conform Never to Identifiable]: <https://forums.swift.org/t/pitch-conform-never-to-identifiable/50110>

**Review by benrimmington — COMMENTED** [review]
As requested, I've added another review, but this pull request also has to be approved by someone from the standard library and/or documentation team.

**Comment by bradleymackey** [comment]
@AnthonyLatsis Yikes this one slipped off my radar! I just recreated my fork for other contributions - I don't see why not though, I'd be happy to work on getting this merged!

**Comment by AnthonyLatsis** [comment]
FWIW, you know that the fork can (and ought to) be reused indefinitely for contributions to the forked repo, don’t you?

> I don't see why not though, I'd be happy to work on getting this merged!

Great! Feel free to ping me if/when you open a new PR.

**Comment by bradleymackey** [comment]
@AnthonyLatsis Yes, it'd just been a while since I'd used it and felt like a "fresh start" 😄. Will do, thanks!

**Pull request closed without merging**